### PR TITLE
Various fixes

### PR DESCRIPTION
**Pull request by nsajko**

### CODE REVIEW

**Comment by nsajko** [comment]
The FreeBSD failure seems unrelated to my change.

File: core/crypto.go [hunk 1]
```diff
@@ -84,16 +84,25 @@ func Hash(b []byte) []byte {
 
 // Scramble fills a given buffer with cryptographically-secure random bytes.
 func Scramble(buf []byte) {
+	// TODO: This should call some assembly routines like Wipe does, for the same
```
Comment (awnumar) [inline core/crypto.go]: What would those assembly routines do? The only option is probably to implement system calls but that's highly platform specific and I'm not sure why some assembly would be less likely to be optimized out.

The compiler likely has some heuristic based on side-effects and some assembly doing the same thing as a standard call would be no different in this regard.

Perhaps there's some way of tricking the compiler into thinking that there's a side effect, by mutating some global or accessing the slice right after.

File: core/crypto.go [hunk 2]
```diff
@@ -84,16 +84,25 @@ func Hash(b []byte) []byte {
...
 	if _, err := rand.Read(buf); err != nil {
 		Panic(err)
 	}
 }
 
+// Unsafely "import" the unexported memclrNoHeapPointers function from runtime. This will
+// break if memclrNoHeapPointers gets renamed or changed.
+//
+//go:linkname memclr runtime.memclrNoHeapPointers
+func memclr(ptr unsafe.Pointer, n uintptr)
```
Comment (awnumar) [inline core/crypto.go]: Interesting, I didn't know you could do this.

File: core/crypto.go [hunk 3]
```diff
@@ -84,16 +84,25 @@ func Hash(b []byte) []byte {
...
+
 // Wipe takes a buffer and wipes it with zeroes.
 func Wipe(buf []byte) {
-	for i := range buf {
-		buf[i] = 0
-	}
+	// Use directly the zeroing function from the Go runtime implemented in assembly,
+	// as a best effort to prevent the zeroization from being optimized out as allowed
+	// by the Go specification.
+	memclr(unsafe.Pointer(&buf[0]), uintptr(len(buf)))
```
[the remark this answers is not in the document]
Comment (nsajko) [reply core/crypto.go]: The idea is that Go implementations are not likely to analyze assembly. Zeroization in native go code thus seems more prone to dead store elimination.

File: core/crypto.go (hunk 1 above)
Comment (nsajko) [reply core/crypto.go]: The hope was that some no-op trick could be done like here: https://boringssl.googlesource.com/boringssl/+/787b26cc51dc5f4b3f423e045bede996038bddd8/crypto/mem.c#156

An alternative would be something like this: https://github.com/golang/go/issues/33325

Comment (awnumar) [reply core/crypto.go]: Referring to discussion in the proposal you opened, what do you think about adding `KeepAlive` calls to these functions instead?

Comment (nsajko) [reply core/crypto.go]: Yeah, appending a KeepAlive call is probably better because of maintainability.

**Comment by awnumar** [comment]
Looks good to me. Add yourself to the [authors](https://github.com/awnumar/memguard/blob/master/AUTHORS) file and I'll merge this.

**Pull request merged**